transmart-api-server/src/main/groovy/org/transmartproject/api/server/user/KeycloakUserResourceService.groovy

```diff
@@ -0,0 +1,136 @@
+import org.transmartproject.core.users.AuthorisationChecks
+import org.transmartproject.core.users.ProtectedOperation
+import org.transmartproject.core.users.ProtectedResource
+import org.transmartproject.core.users.User
+import org.transmartproject.core.users.UsersResource
+
+import java.security.Principal
+import java.text.ParseException
+
+@Component
+@Primary
+@Slf4j
+class KeycloakUserResourceService implements UsersResource {
+
+    @Autowired
+    AuthorisationChecks authorisationChecks
+
+    @Override
+    User getUserFromUsername(String username) throws NoSuchResourceException {
+        throw new UnsupportedOperationException()
+    }
+
+    @Override
+    List<User> getUsers() {
+        throw new UnsupportedOperationException()
+    }
+
+    @Override
+    List<User> getUsersWithEmailSpecified() {
+        throw new UnsupportedOperationException()
+    }
+
+    @Override
+    User getUserFromPrincipal(Principal principal) {
+        assert principal instanceof OAuth2Authentication
+        def authentication = principal.userAuthentication
+        assert authentication: 'User is not authenticated.'
+
+        final String username = principal.name
+        List<String> authorities = principal.authorities*.authority
+        final boolean admin = authorities.remove('ROLE_ADMIN')
+        Multimap<String, ProtectedOperation> accessStudyTokenToOperations =
+                buildStudyTokenToOperationsMultimap(authorities, username)
+
+        def details = authentication.details
+        final String realName
+        final String email
+        if (details instanceof Map) {
+            realName = details.name
+            email = details.email
+        } else {
+            log.warn("Unexpected user details object for ${username} user. Expected map but was ${details}. Hence email and name can't be parsed.")
+            realName = null
+            email = null
+        }
+
+        createUser(username, realName, email, admin, accessStudyTokenToOperations)
+    }
+
+    private User createUser(final String username,
+                            final String realName,
+                            final String email,
+                            final boolean admin,
+                            final Multimap<String, ProtectedOperation> accessStudyTokenToOperations) {
+        new User() {
+            @Override
+            Long getId() {
+                throw new UnsupportedOperationException()
+            }
+
+            @Override
+            String getUsername() {
+                username
+            }
+
+            @Override
+            String getRealName() {
+                realName
+            }
+
+            @Override
+            String getEmail() {
+                email
+            }
+
+            @Override
+            boolean canPerform(ProtectedOperation operation, ProtectedResource protectedResource) {
+                authorisationChecks.canPerform(this, operation, protectedResource)
```

Comment:
Precisely why I think: why would you ask that to the user?

Reply:
That's legacy. The method has been deprecated in the core-api's `User` interface recently.

Reply:
What we could do though. Remove all usages of the method in the code.